### PR TITLE
test: Updated benchmark test results to output result files 

### PR DESCRIPTION
**Pull request by mrickard**
## Description

Updated `run-bench.js` and `benchmark.js` to complete asynchronous tests before generating statistics
Updated `run-bench.js` to allow JSON output to a file, and consolidate printing logic
Reduced cognitive complexity in `run-bench.js`

## How to Test

`npm run bench` to run benchmark tests and output json result files to `benchmark_results`
Given two benchJson outputs, `node bin/compare-bench-results.js {file1.json} {file2.json}`  will compare two benchmark runs, and write a markdown file to `benchmark_results`.

## Related Issues

Closes NR-281645

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -19,3 +19,5 @@ test/versioned-external/TEMP_TESTS
 nr-security-home
 # Needed for testing
 !test/integration/moduleLoading/node_modules
+# benchmark results
+benchmark_*.json
```
Comment (jsumners-nr) [inline .gitignore:23]: Does this mean that the benchmark tests are going to dump results files into the root of the project? If so, can we make them go to a common directory, e.g. `<project_root>/benchmark_results/`?

File: bin/run-bench.js [hunk 2]
```diff
@@ -48,35 +39,37 @@ if (tests.length === 0 && globs.length === 0) {
...
+        resolve()
+      })
+    })
   }
 
-  finish() {
+  async finish() {
+    if (opts.json) {
+      const content = JSON.stringify(this._tests, null, 2)
+      const fileName = `benchmark_${new Date().getTime()}.json`
```
Comment (jsumners-nr) [inline bin/run-bench.js:68]: Is it one file for all benchmark suites per benchmark run? Or does each suite generate its own file?

File: .gitignore (hunk 1 above)
Comment (bizob2828) [reply .gitignore:23]: yea, that's prob a good idea

File: package.json [hunk 3]
```diff
@@ -159,6 +159,7 @@
   },
   "scripts": {
     "bench": "node ./bin/run-bench.js",
+    "benchJson": "node ./bin/run-bench.js --json",
```
Comment (bizob2828) [inline package.json:162]: im not sure i understand the need to have a diff script. can't the `bench` script just output json to a file? I find the previous default, although now enhanced, not very useful as the data just gets written to stdout.

File: bin/run-bench.js (hunk 2 above)
Comment (bizob2828) [reply bin/run-bench.js:68]: it's one per execution

**Review by bizob2828 — CHANGES_REQUESTED** [review]
much better, a lot of great improvements here, just a few suggestions

File: package.json (hunk 3 above)
Comment (mrickard) [reply package.json:162]: I could see wanting to debug by writing to console, but that's going to be the less common use case by far. I'll make output to file the default.

**Review by jsumners-nr — APPROVED** [review]
Looks good to me.

**Review by bizob2828 — APPROVED** [review]
Nice work 👏

**Pull request merged**